rfd/xxxx-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,626 @@
+
+*TODO (Tim):* Per learnings from the demo, generating config may require an
+active API client (auth, proxy, etc). E.g. SSH does to generate `known_hosts`
+and to fetch the proxy address/port/etc. Perhaps it might make sense (where
+it makes sense) to generate these during the renew loop and configure them via
+`tbot.yaml`?
+
+#### Join Script
+
+We intend to develop a bot join script, similar in functionality to the node
+join script that exists in Teleport enterprise. While not required, this will
+make it even easier to install and run the bot as a systemd service on the
+target machine.
+
+TODO: support for auto-join on AWS without token
+
+### Implementation
+
+The bot will be implemented as a standalone tool (ie under `tool/tbot`) and
+_not_ as a service that runs in a `teleport` process.
+
+The bulk of the new logic (parsing configuration, writing certificates to a
+destination, etc.) will be implemented in a new `renew` package, which will
+enable us to start with a standalone bot, and incorporate this functionality
+into other parts of Teleport (ie database access) in the future.
+
+#### API Objects
+
+On the backend, registering a new bot (via `tctl`) creates a set of related API
+objects:
+
+- an object representing the bot itself
+- a non-interactive User for the bot to act as (when user certs are requested)
+- a role for the bot, which will contain the minimum set of permissions
+  necessary to watch for CA rotations, as well as the ability to impersonate
+  other roles
+- a token that the bot can use to join the cluster for the first time
+
+In order to discover related objects and ensure that no "zombie" items are left behind,
+`tctl` will ensure that:
+
+- the objects related to a bot all follow a standardized naming convention that
+  includes the name of the bot
+- a new `teleport.dev/owned-by` label is applied to all resources related to a
+  particular bot
+
+Note: end users should never need to manually create, modify, or delete these
+objects. All operations on bots and their related objects is performed through
+the `tctl bots` subcommand.
+
+#### `tbot` configuration
+
+The goal is to keep tbot configuration to a minimum to make for a quick getting
+started experience, however we recognize many use cases may quickly become
+difficult to encode on the command line.
+
+To that end, we'd like to follow the philosophy of "keep simple things simple
+but make complex things possible": the simplest / most popular use cases should
+ideally require minimal (zero?) configuration beyond a concise CLI, but more
+complicated use cases can make use of a `tbot.yaml` configuration file.
+
+An example `tbot.yaml` with all advanced options specified:
+
+```yaml
+# Storage for the bot's private data
+storage:
+  # can also support any backend supporting both store + fetch
+  directory: /var/lib/tbot
+
+sinks:
+  - directory: /home/alice/.tbot
+    roles: [a, b, c]
+
+  # TODO: probably need to specify permissions (this will be ugly)
```

Comment:
why not supply secure defaults? (bot is only allowed to write, not read)

Reply:
We most likely will not be able to enforce secure defaults from the bot's process (unless it runs as root, which I'd prefer to avoid). I'll remove the TODO here since it just adds confusion.

This relates back to filesystem permissions from before, if we go the ACL route users would need to create the files using their target Unix user and then grant the bot's Unix user write-only access via ACLs (we could provide a helper subcommand for this). There's still more research to do here, I intend test this out and will update the doc.